### ISSUE TITLE
Macros

### ISSUE THREAD

**Issue by tommyiswhoiam** [Orange-OpenSource/hurl#1273]
### Problem to solve

There is a need for a more efficient way to handle JSON-RPC requests in Hurl. Currently, each request has to be written out in full, leading to duplicated code and longer scripts.

### Proposal

Introduce macros to Hurl that allow users to define reusable blocks of code. In this particular case, we propose a `RPC` macro that will handle JSON-RPC requests. This macro will take two parameters: `method` and `params`, which will be used to specify the method to be called and the parameters to be passed to it. The macro will handle the construction of the JSON-RPC request and the parsing of the response.

Here's the example definition of the `RPC` macro:

```ini
MACRO RPC method params
POST {{endpoint}}
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": {{method}},
  "params": {{params}}
}

HTTP 200
ENDMACRO
```

Here's an example usage of the macro:

```ini
RPC "sum" [2, 2]
[Asserts]
jsonpath "$.result" == 4
```

And here's the code that the macro will compile to:

```ini
POST {{endpoint}}
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "sum",
  "params": [2, 2]
}

HTTP 200
[Asserts]
jsonpath "$.result" == 4
```

Other example with positional parameters:
```ini
MACRO RPC
POST {{endpoint}}
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": {{$1}},
  "params": {{$2}}
}

HTTP 200
ENDMACRO
```

### Additional context and resources

Macros are a common feature in many other scripting languages, and they have been shown to improve code readability and reduce duplicated code. By introducing macros to Hurl, users will be able to write more efficient and maintainable scripts.

### Tasks to complete

- [ ] Add support for macros to the Hurl library
- [ ] Update the documentation to reflect the new feature
- [ ] Add tests to ensure the macro works as expected
- [ ] Release a new version of the library with the changes.

**Comment by tommyiswhoiam** [comment]
In combination with imports, this feature can address a significant number of user requests.

**Comment by rbas** [comment]
I really value how hurl simple is. But indeed if you would like to use it for test several use cases of your API you're ending up with duplicated code. Some way to reduce duplications will be really good either Macros, Imports, or both.

**Comment by jcamiel** [comment]
Hi @tommyiswhoiam @rbas thanks for the feedbacks. I would like juste to warn you that we're going to close this one as 'won't fix', and I really want to explain the reasoning. We heard numerous feedback about import, reuse (see #115 #317), we know that the current way of working with Hurl can be seen as limiting. 

Currently, we try to be focused, and, as Hurl is a side and passion project, we would like to work on features that are important for us. 
Now, we would like to:
- improve the tooling story, with the objective of having a very good support for VSCode and IntelliJ
- work on running tests in parallel (see our nearly first issues #87 and #88)

These two features are going to be our priority for this year.

Reusing, importing, macros are also important but it's not our current priority. These are good features, but they come with _really major additional complexity_ that we don't want to support, for the moment. We've designed the Hurl file format to be like a Markdown for HTTP rather than a proper programming language. This, for sure, is limiting, but we also see it as a force of the tool. With a simple format, you can image simple way of generating Hurl files (with template engine, scripts etc...). I hope this answer is not too deceptive (and closing it now doesn't mean we won't revisit it in the future)

**Comment by rbas** [comment]
Hi @jcamiel thank you for your answer. I fully understand your decision. The fact that you've designed hurl so, is big power. I'm looking for a tool to cover different use cases for API tests. Hurt is nicely declarative language readable for most people. I think that is a good achievement.
But if you would like to cover different use cases you will write some code over and over, which is not good. 

I understand that imports or macros are not priorities for you for this year. Would you mind if I will use my spare free time for it?

**Comment by jcamiel** [comment]
Hi @rabs, you can spare you free time as you want of course 😅! That said, we won't be obligated to merge any work as we've tried to be clear in the [CONTRIBUTING guide](https://github.com/Orange-OpenSource/hurl/blob/master/CONTRIBUTING.md). If you need some help because the code is not clear enough, don't hesitate to ask us, we try to improve the current code documentation.
